### PR TITLE
Feature/1487 paginator component

### PR DESCRIPTION
**Pull request by anna-lach**
Contains:
#1487 
#1486 

Fixes:
#1617

### CODE REVIEW

**Comment by Diaan** [comment]
the (keyboard) focus is lost when you select an option from the overflow menu with the enter key. Nothing has the focus once the menu closes.

File: packages/components/paginator/src/paginator.ts [hunk 1]
```diff
@@ -0,0 +1,460 @@
...
+      this.activePage = Number(target.value);
+    } else {
+      this.activePage = Number(target.innerText?.trim());
+    }
+
+    this.#setCurrentlyVisibleItems();
+    this.#update();
+  }
+
+  #update(): void {
```
[the remark this answers is not in the document]
Comment (jpzwarte) [reply packages/components/paginator/src/paginator.ts:277]: Also, `this.#update()` feels too similar to `this.update()` (Lit method). Please provide a more specific name.

File: packages/components/paginator/src/page.ts [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+
+declare global {
+  interface HTMLElementTagNameMap {
+    'sl-page': Page;
+  }
+}
+
+/**
+ * A page component that can be used as part of the paginator, representing pages in the paginator.
+ *
```
Comment (jpzwarte) [inline packages/components/paginator/src/page.ts:12]: Remove empty line plz

Comment (jpzwarte) [reply packages/components/paginator/src/page.ts:12]: Perhaps add here that this is an internal component and not a public one?

File: packages/components/paginator/src/page.ts [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+ */
+export class Page extends LitElement {
+  /** @internal */
+  static override styles: CSSResultGroup = styles;
+
+  /** @internal Page number used for the aria-label. */
+  #pageNumber = '';
```
Comment (jpzwarte) [inline packages/components/paginator/src/page.ts:19]: Why not do `@state() pageNumber?: string;`? That way, when `slotchange` event fires, it automatically rerenders the template.

File: packages/components/paginator/src/page.ts [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+
+  override connectedCallback(): void {
+    super.connectedCallback();
+
+    this.#pageNumber = Array.from(this.childNodes)
```
Comment (jpzwarte) [inline packages/components/paginator/src/page.ts:24]: Is this really necessary? Isn't this code already in `#onSlotChange`?

File: packages/components/paginator/src/paginator.stories.ts [hunk 5]
```diff
@@ -0,0 +1,152 @@
...
+      ></sl-paginator>
+    `;
+  }
+} satisfies Meta<Props>;
+
+export const Basic: Story = {};
+
+export const Overflow: Story = {
+  args: {
+    ...Basic.args,
```
Comment (jpzwarte) [inline packages/components/paginator/src/paginator.stories.ts:57]: You can remove `...Basic.args` everywhere, since `Basic` doesn't define any custom args :)

File: packages/components/paginator/src/paginator.stories.ts [hunk 6]
```diff
@@ -0,0 +1,152 @@
...
+    total: 900
+  }
+};
+
+export const Mobile: Story = {
+  args: {
+    ...Basic.args,
+    activePage: 5
+  },
+  parameters: {
```
Comment (jpzwarte) [inline packages/components/paginator/src/paginator.stories.ts:67]: Nitpicking, but i would put `parameters` above `args`, because `args` and `render` are more "connected".

File: packages/components/paginator/src/paginator.stories.ts [hunk 7]
```diff
@@ -0,0 +1,152 @@
...
+    `;
+  }
+};
+
+export const PageSizeComponent: Story = {
+  args: {
+    ...Basic.args
+  },
+  render: ({ pageSizes, itemsPerPage }) => {
+    return html`<sl-page-size .pageSizes=${pageSizes} .itemsPerPage=${itemsPerPage}></sl-page-size>`;
```
Comment (jpzwarte) [inline packages/components/paginator/src/paginator.stories.ts:89]: I would create separate `page-size.stories.ts` files etc. for separate components. You could nest them inside the `Navigation/Paginator` in the sidebar though: `title: 'Navigation/Paginator/PageSize`

File: packages/components/paginator/src/items-counter.ts [hunk 8]
```diff
@@ -0,0 +1,95 @@
...
+  @property({ type: Number, attribute: 'active-page' }) activePage = 1;
+
+  /** @internal Currently visible items on the current page. */
+  @state() currentlyVisibleItems = 1;
+
+  /** Items per page, if not set - default to 10. */
+  @property({ type: Number, attribute: 'items-per-page' }) itemsPerPage = 10;
+
+  /** @internal Pages amount. */
+  #pages = 1;
```
Comment (jpzwarte) [inline packages/components/paginator/src/items-counter.ts:31]: Please put private variables, *after* static variables and before non-private vars :)

File: packages/components/paginator/src/items-counter.ts [hunk 9]
```diff
@@ -0,0 +1,95 @@
...
+
+  /** Total amount of items, if not set - default to 1. */
+  @property() total = 1;
```
Comment (jpzwarte) [inline packages/components/paginator/src/items-counter.ts:34]: Is this a public property? If so, please add `type: Number`; otherwise, please add `@internal` to the comment

File: packages/components/paginator/src/paginator.ts [hunk 10]
```diff
@@ -0,0 +1,460 @@
...
+      });
+    }
+
+    this.#firstUpdate = false;
+  }
+
+  override render(): TemplateResult {
+    return html`
+      <nav class="container">
+        <sl-button
```
[the remark this answers is not in the document]
Comment (anna-lach) [reply packages/components/paginator/src/paginator.ts:183]: Is it really necessary? Anna Karoń was also testing it, and she didn't find any issue here connected with those arias missing?

File: packages/components/paginator/src/paginator.ts [hunk 11]
```diff
@@ -0,0 +1,460 @@
...
+
+    if (this.activePage === this.#pages) {
+      const itemsOnLastPage = this.total % this.itemsPerPage;
+      this.currentlyVisibleItems = itemsOnLastPage === 0 ? this.itemsPerPage : itemsOnLastPage;
+    } else {
+      this.currentlyVisibleItems = this.itemsPerPage;
+    }
+  }
+
+  #createMenuButton(): HTMLLIElement {
```
[the remark this answers is not in the document]
Comment (anna-lach) [reply packages/components/paginator/src/paginator.ts:414]: Here we are adding the pages:

image


And the menu button/s should be somewhere in between the pages, depending on a situation, that's why I've added it here.

File: packages/components/paginator/src/paginator-page.ts [hunk 12]
```diff
@@ -0,0 +1,36 @@
...
+
+/**
+ * A page internal component, that is used as part of the paginator, representing pages in the paginator.
+ */
+export class PaginatorPage extends LitElement {
+  /** @internal */
+  static override styles: CSSResultGroup = styles;
+
+  /** @internal Page number used for the aria-label. */
+  @state() pageNumber?: string;
```
Comment (jpzwarte) [inline packages/components/paginator/src/paginator-page.ts:19]: If we use `page` in the datasource, perhaps use the same name here?

File: packages/components/paginator/src/paginator-size.ts [hunk 13]
```diff
@@ -0,0 +1,119 @@
...
+@localized()
+export class PaginatorSize extends ScopedElementsMixin(LitElement) {
+  /** @internal */
+  static override styles: CSSResultGroup = styles;
+
+  /** Provided data source. */
+  @property({ attribute: false }) dataSource?: DataSource;
+
+  /** Items per page. Default to the first item of pageSizes, if pageSizes is not set - default to 10. */
+  @property({ type: Number, attribute: 'items-per-page' }) itemsPerPage?: number;
```
Comment (jpzwarte) [inline packages/components/paginator/src/paginator-size.ts:34]: We use `pageSize` in datasource. Why not use the same name here?

File: packages/components/paginator/src/paginator-size.ts [hunk 14]
```diff
@@ -0,0 +1,119 @@
...
+
+  /** @internal Emits when the page size has been selected/changed. */
+  @event({ name: 'sl-page-size-change' }) pageSizeChangeEvent!: EventEmitter<SlChangeEvent<number>>; //EventEmitter<SlPageSizeChangeEvent>;
+
+  /** Page sizes - array of possible page sizes e.g. [5, 10, 15]. */
+  @property({ type: Number, attribute: 'page-sizes' }) pageSizes?: number[];
+
+  /** @internal */
+  static get scopedElements(): ScopedElementsMap {
```
Comment (jpzwarte) [inline packages/components/paginator/src/paginator-size.ts:43]: Please move static code above non-static code

File: packages/components/paginator/src/paginator-size.ts [hunk 15]
```diff
@@ -0,0 +1,119 @@
...
+  static override styles: CSSResultGroup = styles;
+
+  /** Provided data source. */
+  @property({ attribute: false }) dataSource?: DataSource;
+
+  /** Items per page. Default to the first item of pageSizes, if pageSizes is not set - default to 10. */
+  @property({ type: Number, attribute: 'items-per-page' }) itemsPerPage?: number;
+
+  /** @internal Emits when the page size has been selected/changed. */
+  @event({ name: 'sl-page-size-change' }) pageSizeChangeEvent!: EventEmitter<SlChangeEvent<number>>; //EventEmitter<SlPageSizeChangeEvent>;
```
Comment (jpzwarte) [inline packages/components/paginator/src/paginator-size.ts:37]: Remove comment plz

File: packages/components/paginator/src/paginator-size.ts [hunk 16]
```diff
@@ -0,0 +1,119 @@
...
+    };
+  }
+
+  override disconnectedCallback(): void {
+    this.dataSource?.removeEventListener('sl-update', this.#onUpdate);
+
+    super.disconnectedCallback();
+  }
+
+  override firstUpdated(changes: PropertyValues<this>): void {
```
Comment (jpzwarte) [inline packages/components/paginator/src/paginator-size.ts:56]: Do we want to only run this on `firstUpdated`? Or everytime `dataSource` or `itemsPerPage` changes? The latter i assume? So rename `firstUpdated` to `willUpdate`?

File: packages/components/paginator/src/paginator-size.ts [hunk 17]
```diff
@@ -0,0 +1,119 @@
...
+    }
+
+    if (this.dataSource) {
+      this.dataSource.addEventListener('sl-update', this.#onUpdate);
+    }
+  }
+
+  override render(): TemplateResult {
+    return html`
+      <span>${msg('Items per page')}:</span>
```
Comment (jpzwarte) [inline packages/components/paginator/src/paginator-size.ts:70]: If you use `<sl-label>` here, you don't need to set the `aria-label` attribute on `<sl-select>`?

File: packages/components/paginator/src/paginator-size.ts [hunk 18]
```diff
@@ -0,0 +1,119 @@
...
+              aria-label=${`${this.itemsPerPage} ${msg('Items per page')}`}
+              size="lg"
+              value=${this.itemsPerPage}
+            >
+              ${this.pageSizes.map(
+                size => html`
+                  <sl-select-option
+                    aria-label=${`${size} ${msg('Items per page')}`}
+                    @click=${this.#setValue}
+                    value=${size}
```
Comment (jpzwarte) [inline packages/components/paginator/src/paginator-size.ts:83]: `.value=${size}` (missing the dot)

File: packages/components/paginator/src/paginator-size.ts [hunk 19]
```diff
@@ -0,0 +1,119 @@
...
+            <sl-select
+              aria-label=${`${this.itemsPerPage} ${msg('Items per page')}`}
+              size="lg"
+              value=${this.itemsPerPage}
+            >
+              ${this.pageSizes.map(
+                size => html`
+                  <sl-select-option
+                    aria-label=${`${size} ${msg('Items per page')}`}
+                    @click=${this.#setValue}
```
Comment (jpzwarte) [inline packages/components/paginator/src/paginator-size.ts:82]: Isn't using `@sl-change=${this.#onChange}` on `<sl-select>` better?

File: packages/components/paginator/src/paginator-size.ts [hunk 20]
```diff
@@ -0,0 +1,119 @@
...
+                  >
+                    ${size}
+                  </sl-select-option>
+                `
+              )}
+            </sl-select>
+          `
+        : nothing}
+      <!-- We want this to be read every time the active page changes. -->
+      <span id="live" aria-live="polite" aria-atomic="true">
```
Comment (jpzwarte) [inline packages/components/paginator/src/paginator-size.ts:93]: Why span instead of div?

File: packages/components/paginator/src/paginator-size.ts [hunk 21]
```diff
@@ -0,0 +1,119 @@
...
+
+  #setValue(event: Event): void {
+    const newValue = Number((event.target as SelectOption).value);
+    if (this.itemsPerPage !== newValue) {
+      this.itemsPerPage = newValue;
+
+      /** Emits amount of selected items per page */
+      this.pageSizeChangeEvent.emit(newValue);
+
+      if (this.dataSource) {
```
Comment (jpzwarte) [inline packages/components/paginator/src/paginator-size.ts:107]: if statement isn't necessary. You're already checking for this using `?`: `this.dataSource?.setPageSize()`

File: packages/components/paginator/src/paginator-status.scss [hunk 22]
```diff
@@ -0,0 +1,16 @@
...
+  font: var(--sl-text-new-body-md);
+  font-weight: var(--sl-text-typeset-font-weight-icon-regular);
+  padding: 0;
+  text-wrap: nowrap;
+}
+
+/** Hiding the aria-live element for the UI, but not for the screen readers. */
+#live {
+  block-size: 0;
+  display: block;
```
Comment (jpzwarte) [inline packages/components/paginator/src/paginator-status.scss:13]: If you use a `<div>`, then you can remove this line.

File: packages/components/paginator/src/paginator-status.spec.ts [hunk 23]
```diff
@@ -0,0 +1,218 @@
...
+import { html } from 'lit';
+import '../register.js';
+import { PaginatorStatus } from './paginator-status.js';
+
+describe('sl-paginator-status', () => {
+  let el: PaginatorStatus;
+
+  describe('defaults', () => {
+    beforeEach(async () => {
+      el = await fixture(html` <sl-paginator-status></sl-paginator-status> `);
```
Comment (jpzwarte) [inline packages/components/paginator/src/paginator-status.spec.ts:13]: Unnecessary whitespace :)

File: packages/components/paginator/src/paginator-status.spec.ts [hunk 24]
```diff
@@ -0,0 +1,218 @@
...
+
+      await new Promise(resolve => setTimeout(resolve, 100));
```
Comment (jpzwarte) [inline packages/components/paginator/src/paginator-status.spec.ts:15]: Please add a comment explaining why the `setTimeout` is necessary.

File: packages/components/paginator/src/paginator-status.ts [hunk 25]
```diff
@@ -0,0 +1,113 @@
...
+@localized()
+export class PaginatorStatus extends LitElement {
+  /** @internal */
+  static override styles: CSSResultGroup = styles;
+
+  /** @internal Pages amount. */
+  #pages = 1;
+
+  /** Currently active page, if not set - default to 1. */
+  @property({ type: Number, attribute: 'active-page' }) activePage = 1;
```
Comment (jpzwarte) [inline packages/components/paginator/src/paginator-status.ts:26]: Why `activePage` instead of just `page`? (also matches datasource)

File: packages/components/paginator/src/paginator-status.ts [hunk 26]
```diff
@@ -0,0 +1,113 @@
...
+
+  /** @internal Currently visible items on the current page. */
+  @state() currentlyVisibleItems = 1;
+
+  /** Provided data source. */
+  @property({ attribute: false }) dataSource?: DataSource;
+
+  /** Items per page, if not set - default to 10. */
+  @property({ type: Number, attribute: 'items-per-page' }) itemsPerPage = 10;
```
Comment (jpzwarte) [inline packages/components/paginator/src/paginator-status.ts:35]: `pageSize`

File: packages/components/paginator/src/paginator-status.ts [hunk 27]
```diff
@@ -0,0 +1,113 @@
...
+
+  /** Total amount of items, if not set - default to 1. */
+  @property({ type: Number }) total = 1;
```
Comment (jpzwarte) [inline packages/components/paginator/src/paginator-status.ts:38]: `pageCount`

File: packages/components/paginator/src/paginator-status.ts [hunk 28]
```diff
@@ -0,0 +1,113 @@
...
+
+  override firstUpdated(changes: PropertyValues<this>): void {
+    super.firstUpdated(changes);
+
+    this.#pages = Math.ceil(this.total / this.itemsPerPage);
+
+    this.#setCurrentlyVisibleItems();
+
+    if (this.dataSource) {
+      this.dataSource.addEventListener('sl-update', this.#onUpdate);
```
Comment (jpzwarte) [inline packages/components/paginator/src/paginator-status.ts:54]: Move to `updated()`

File: packages/components/paginator/src/paginator.ts [hunk 29]
```diff
@@ -0,0 +1,531 @@
...
+
+  /** @internal Pages amount. */
+  #pages = 1;
+
+  /** Observe changes in size of the container. */
+  #observer = new ResizeObserver(() => {
+    requestAnimationFrame(() => this.#updateVisibility());
+  });
+
+  get activePage(): number {
```
Comment (jpzwarte) [inline packages/components/paginator/src/paginator.ts:68]: `page` instead of `activePage`

**Review by jpzwarte — APPROVED** [review]

File: packages/components/paginator/src/paginator-status.ts (hunk 27 above)
Comment (jpzwarte) [reply packages/components/paginator/src/paginator-status.ts:38]: So i read the comment wrong: `totalItems` better reflects the property i think. And also makes clear it's unrelated to pageCount

File: packages/components/paginator/src/paginator.ts [hunk 30]
```diff
@@ -46,61 +73,75 @@ export class Paginator extends ScopedElementsMixin(LitElement) {
...
+      case 'xs':
+        return 6;
+      case 'sm':
+        return 7;
+      case 'md':
+        return 9;
+      case 'lg':
+        return 11;
+      default:
+        return 11;
```
Comment (Diaan) [inline packages/components/paginator/src/paginator.ts:119]: you can remove `case 'lg'`, since it's the same as the default

File: packages/locales/src/nl.ts [hunk 31]
```diff
@@ -29,17 +26,20 @@ export const templates = {
   s62aa2ba3fe47ece4: 'Vink tenminste één optie aan.',
   s6580790b036f0c6f: 'actief',
   s6abb1cd87fe0114e: 'Home',
-  s708f844ea6d35edd: str`Momenteel actief pagina ${0} van ${1} pagina's`,
+  s6ef3b70aa370dba4: str`Aantal items per pagina: ${0}`,
+  s7d09afc329d9995b: str`Pagina ${0} van ${1}`,
   s8079b71872346425: 'Voer een waarde in.',
   s864c948d4629240f: 'Toon alles',
   s886db4d476f66522: 'Geselecteerde opties',
+  s88b887b46242ba96: str`Ga naar vorige pagina ${0}`,
```
Comment (Diaan) [inline packages/locales/src/nl.ts:34]:
```suggestion
  s88b887b46242ba96: str`Ga naar vorige pagina (${0})`,
```

File: packages/locales/src/nl.ts [hunk 32]
```diff
@@ -29,17 +26,20 @@ export const templates = {
...
   s8c1f47dc13ead0c3: str`Maak "${0}" aan`,
+  s9ad43731ebd435aa: str`Momenteel worden ${0} tot ${1} van ${2} weergegeven`,
   s9cd91b99a108a45f: 'Lijst met verborgen elementen',
   s9f7452dd75d54d31: 'fout',
   sa447e2f801a7eb2d: 'Kies een optie.',
+  sa477d2feda7d475e: str`Ga naar de volgende pagina ${0}`,
```
Comment (Diaan) [inline packages/locales/src/nl.ts:40]:
```suggestion
  sa477d2feda7d475e: str`Ga naar de volgende pagina (${0})`,
```

File: packages/components/data-source/src/array-data-source.ts [hunk 33]
```diff
@@ -10,6 +10,7 @@ import {
 export class ArrayDataSource<T = any> extends DataSource<T> {
```
[the remark this answers is not in the document]
Comment (anna-lach) [reply packages/components/data-source/src/array-data-source.ts]: Will be done in a separate issue: https://github.com/sl-design-system/components/issues/1625

**Review by Diaan — APPROVED** [review]
